### PR TITLE
Feature/adc1 dma support

### PR DESCRIPTION
**Pull request by aborowski-msol**
ADC1 DMA support with multiple channels reading.

### CODE REVIEW

File: drivers/adc/adc_stm32.c [hunk 1]
```diff
@@ -233,6 +260,94 @@ struct adc_stm32_cfg {
...
+	ADC_TypeDef *adc = (ADC_TypeDef *)config->base;
+
+	SCB_InvalidateDCache_by_Addr(data->buffer, sizeof(uint16_t) * data->channel_count);
+	SCB_InvalidateDCache_by_Addr(data->repeat_buffer, sizeof(uint16_t) * data->channel_count);
+
+	adc_context_on_sampling_done(&data->ctx, dev);
+}
+
+/* Setup DMA for single or multiple read */
+/* Could probably be done once, but the length will change depending on sequence */
```
Comment (heinwessels) [inline drivers/adc/adc_stm32.c]: Should be removed.

File: drivers/adc/adc_stm32.c [hunk 2]
```diff
@@ -233,6 +260,94 @@ struct adc_stm32_cfg {
...
+
+	/* give the fifo mode from the DT */
+	blk_cfg->fifo_mode_control = data->dma.fifo_threshold;
+
+	/* direction is given by the DT */
+	stream->dma_cfg.head_block = blk_cfg;
+	stream->dma_cfg.user_data = data;
+
+	/* pass our client origin to the dma: data->dma.channel */
+	LOG_DBG("Ready to do config");
```
Comment (heinwessels) [inline drivers/adc/adc_stm32.c]: Remove

File: drivers/adc/adc_stm32.c [hunk 3]
```diff
@@ -233,6 +260,94 @@ struct adc_stm32_cfg {
...
+			LL_ADC_REG_DMA_TRANSFER_LIMITED <<  ADC3_CFGR_DMACFG_Pos);
+		LL_ADC_EnableDMAReq(adc);
+	} else {
+		LL_ADC_REG_SetDataTransferMode(adc, LL_ADC_REG_DMA_TRANSFER_LIMITED);
+	}
+#else
+	LL_ADC_REG_SetDataTransferMode(adc, LL_ADC_REG_DMA_TRANSFER_LIMITED);
+#endif
+
+	LOG_DBG("DMA Configured");
```
Comment (heinwessels) [inline drivers/adc/adc_stm32.c]: Remove

File: drivers/adc/adc_stm32.c [hunk 4]
```diff
@@ -471,39 +586,41 @@ static int start_read(const struct device *dev,
...
-	defined(CONFIG_SOC_SERIES_STM32L0X)
-	LL_ADC_REG_SetSequencerChannels(adc, channel);
+			defined(CONFIG_SOC_SERIES_STM32L0X)
+			LL_ADC_REG_SetSequencerChannels(adc, channel);
 #elif defined(CONFIG_SOC_SERIES_STM32G0X)
-	/* STM32G0 in "not fully configurable" sequencer mode */
-	LL_ADC_REG_SetSequencerChannels(adc, channel);
-	while (LL_ADC_IsActiveFlag_CCRDY(adc) == 0) {
-	}
+			*STM32G0 in "not fully configurable" sequencer mode * /
```
Comment (heinwessels) [inline drivers/adc/adc_stm32.c]: This line is malformed.

File: drivers/adc/adc_stm32.c [hunk 5]
```diff
@@ -651,9 +773,11 @@ static void adc_stm32_isr(const struct device *dev)
...
 
-	*data->buffer++ = LL_ADC_REG_ReadConversionData32(adc);
+	data->buffer[data->samples_count++] = LL_ADC_REG_ReadConversionData32(adc);
 
-	adc_context_on_sampling_done(&data->ctx, dev);
+	if (data->samples_count == data->channel_count) {
+		adc_context_on_sampling_done(&data->ctx, dev);
+	}
 
 	LOG_DBG("ISR triggered.");
```
Comment (heinwessels) [inline drivers/adc/adc_stm32.c:796]: Remove

File: drivers/adc/adc_stm32.c [hunk 6]
```diff
@@ -664,6 +788,8 @@ static int adc_stm32_read(const struct device *dev,
 	struct adc_stm32_data *data = dev->data;
 	int error;
 
+	LOG_DBG("adc_stm32_read");
+
```
Comment (heinwessels) [inline drivers/adc/adc_stm32.c]: Remove

File: drivers/adc/adc_stm32.c [hunk 7]
```diff
@@ -937,41 +1072,127 @@ static const struct adc_driver_api api_stm32_driver_api = {
...
+										\
+	static void adc_stm32_cfg_func_##index(void)				\
+	{									\
+		IRQ_CONNECT(DT_INST_IRQN(index),				\
+			    DT_INST_IRQ(index, priority),			\
+			    adc_stm32_isr, DEVICE_DT_INST_GET(index), 0);	\
+		irq_enable(DT_INST_IRQN(index));				\
+	}
+
+#endif /* CONFIG_ADC_STM32_DMA */
```
Comment (heinwessels) [inline drivers/adc/adc_stm32.c]: These two `STM32_ADC_INIT` functions should be consolidated into one with `#ifdef CONFIG_ADC_STM32_DMA` at the correct locations. I simply copy-and-pasted it to start with just to get something running.

File: drivers/dma/dma_stm32.c [hunk 8]
```diff
@@ -87,6 +87,8 @@ static void dma_stm32_irq_handler(const struct device *dev, uint32_t id)
 	struct dma_stm32_stream *stream;
 	uint32_t callback_arg;
 
+	LOG_DBG("dma_stm32_irq_handler");
```
Comment (heinwessels) [inline drivers/dma/dma_stm32.c]: This file is full of debug logging which should be removed. It was only added for testing.

File: drivers/dma/dma_stm32.c [hunk 9]
```diff
@@ -420,14 +431,27 @@ DMA_STM32_EXPORT_API int dma_stm32_configure(const struct device *dev,
...
+		break;
+	case LL_DMA_PDATAALIGN_HALFWORD:
+		LOG_DBG("\tPeriphOrM2MSrcDataSize LL_DMA_PDATAALIGN_HALFWORD");
+		break;
+	case LL_DMA_PDATAALIGN_WORD:
+		LOG_DBG("\tPeriphOrM2MSrcDataSize LL_DMA_PDATAALIGN_WORD");
+		break;
+	default:
+		LOG_DBG("\tPeriphOrM2MSrcDataSize OOPS");
+	}
```
Comment (heinwessels) [inline drivers/dma/dma_stm32.c]: Remove

File: samples/drivers/adc/boards/nucleo_h743zi.overlay [hunk 10]
```diff
@@ -0,0 +1,20 @@
+&adc1 {
+	/* The <channel> doesn't depend on peripheral */
+	/* The <slot> is the request line, so maybe adc1_dma? */
```
Comment (heinwessels) [inline samples/drivers/adc/boards/nucleo_h743zi.overlay]: `The <slot> is the request line, so maybe adc1_dma?` should be changed.

File: drivers/adc/adc_stm32.c [hunk 11]
```diff
@@ -233,6 +260,94 @@ struct adc_stm32_cfg {
...
+{
+	/* arg directly holds the adc device */
+	struct adc_stm32_data *data = arg;
+
+	/* Read the ADC after DMA says it's done */
+	struct device *adc_dev = (ADC_TypeDef *)data->dev;
+	const struct adc_stm32_cfg *config = adc_dev->config;
+	ADC_TypeDef *adc = (ADC_TypeDef *)config->base;
+
+	SCB_InvalidateDCache_by_Addr(data->buffer, sizeof(uint16_t) * data->channel_count);
```
Comment (heinwessels) [inline drivers/adc/adc_stm32.c]: A comment here about why this is required would be good I think

File: samples/drivers/adc/prj.conf [hunk 12]
```diff
@@ -1 +1,3 @@
 CONFIG_ADC=y
+CONFIG_ADC_ASYNC=y
+CONFIG_ADC_STM32_DMA=y
```
Comment (heinwessels) [inline samples/drivers/adc/prj.conf]: This should not be here. This will attempt to apply `STM32_DMA` for non-STM boards trying to run this sample.

File: samples/drivers/adc/src/main.c [hunk 13]
```diff
@@ -5,42 +5,26 @@
...
-#define ADC_REFERENCE		ADC_REF_INTERNAL
-#define ADC_ACQUISITION_TIME	ADC_ACQ_TIME_DEFAULT
-
-/* Get the numbers of up to two channels */
-static uint8_t channel_ids[ADC_NUM_CHANNELS] = {
-	DT_IO_CHANNELS_INPUT_BY_IDX(DT_PATH(zephyr_user), 0),
-#if ADC_NUM_CHANNELS == 2
-	DT_IO_CHANNELS_INPUT_BY_IDX(DT_PATH(zephyr_user), 1)
-#endif
-};
```
Comment (heinwessels) [inline samples/drivers/adc/src/main.c]: Why did you remove this? We shouldn't change how the sample works, and it will likely break the other samples. The sample doesn't need to reflect how we will implement it. I think you can setup the channels as the sample did it before, and just assume that all channels are from ADC1.

File: samples/drivers/adc/prj.conf [hunk 14]
```diff
@@ -1 +1,3 @@
 CONFIG_ADC=y
+CONFIG_ADC_ASYNC=y
```
Comment (heinwessels) [inline samples/drivers/adc/prj.conf]: `CONFIG_ADC_ASYNC` is not used in the sample application.

File: samples/drivers/adc/boards/nucleo_h743zi.conf [hunk 15]
```diff
@@ -0,0 +1,9 @@
+CONFIG_ADC_STM32_DMA=y
```
Comment (heinwessels) [inline samples/drivers/adc/boards/nucleo_h743zi.conf:1]: The sample application doesn't work without `DMA` when more than one channel is selected. The last ADC ISR isn't triggered for some reason. 

With one channel it completes, but I'm fairly sure if it's returning the ADC data. It stays zero, even though it seems `data->buffer` is filled with good information.

File: samples/drivers/adc/src/main.c [hunk 16]
```diff
@@ -52,19 +36,46 @@ struct adc_channel_cfg channel_cfg = {
...
+
+		/*
+		 * Convert raw reading to millivolts if driver
+		 * supports reading of ADC reference voltage
+		 */
+		int32_t mv_value = raw_value;
+
+		adc_raw_to_millivolts(adc_ref_internal(dev), ADC_GAIN,
+				      ADC_RESOLUTION, &mv_value);
+		LOG_ERR("ADC ch%d reading %d mV", i, mv_value);
```
Comment (heinwessels) [inline samples/drivers/adc/src/main.c]: We should move back to `printk` as the sample used originally.
```suggestion
		printk("ADC ch%d reading %d mV", i, mv_value);
```

File: drivers/adc/adc_stm32.c [hunk 17]
```diff
@@ -233,6 +260,94 @@ struct adc_stm32_cfg {
...
+static void dma_callback(const struct device *dev, void *arg,
+			 uint32_t channel, int status)
+{
+	/* arg directly holds the adc device */
+	struct adc_stm32_data *data = arg;
+
+	/* Read the ADC after DMA says it's done */
+	struct device *adc_dev = (ADC_TypeDef *)data->dev;
+	const struct adc_stm32_cfg *config = adc_dev->config;
+	ADC_TypeDef *adc = (ADC_TypeDef *)config->base;
```
Comment (heinwessels) [inline drivers/adc/adc_stm32.c]: Unused.

**Review by heinwessels — CHANGES_REQUESTED** [review]

File: drivers/adc/adc_stm32.c (hunk 2 above)
Comment (aborowski-msol) [reply drivers/adc/adc_stm32.c]: removed

File: drivers/adc/adc_stm32.c (hunk 1 above)
Comment (aborowski-msol) [reply drivers/adc/adc_stm32.c]: removed

File: drivers/adc/adc_stm32.c (hunk 11 above)
Comment (aborowski-msol) [reply drivers/adc/adc_stm32.c]: added comment

File: drivers/adc/adc_stm32.c (hunk 17 above)
Comment (aborowski-msol) [reply drivers/adc/adc_stm32.c]: removed

File: drivers/adc/adc_stm32.c (hunk 3 above)
Comment (aborowski-msol) [reply drivers/adc/adc_stm32.c]: removed

File: drivers/adc/adc_stm32.c (hunk 4 above)
Comment (aborowski-msol) [reply drivers/adc/adc_stm32.c]: fixed

File: drivers/adc/adc_stm32.c (hunk 5 above)
Comment (aborowski-msol) [reply drivers/adc/adc_stm32.c:796]: removed

File: drivers/adc/adc_stm32.c (hunk 6 above)
Comment (aborowski-msol) [reply drivers/adc/adc_stm32.c]: removed

File: drivers/adc/adc_stm32.c (hunk 7 above)
Comment (aborowski-msol) [reply drivers/adc/adc_stm32.c]: consolidated

File: drivers/dma/dma_stm32.c (hunk 8 above)
Comment (aborowski-msol) [reply drivers/dma/dma_stm32.c]: removed

File: drivers/dma/dma_stm32.c (hunk 9 above)
Comment (aborowski-msol) [reply drivers/dma/dma_stm32.c]: removed

File: samples/drivers/adc/boards/nucleo_h743zi.conf (hunk 15 above)
Comment (aborowski-msol) [reply samples/drivers/adc/boards/nucleo_h743zi.conf:1]: reverted sample app changes

File: samples/drivers/adc/prj.conf (hunk 14 above)
Comment (aborowski-msol) [reply samples/drivers/adc/prj.conf]: removed

File: samples/drivers/adc/prj.conf (hunk 12 above)
Comment (aborowski-msol) [reply samples/drivers/adc/prj.conf]: removed

File: samples/drivers/adc/src/main.c (hunk 13 above)
Comment (aborowski-msol) [reply samples/drivers/adc/src/main.c]: restored

File: samples/drivers/adc/src/main.c (hunk 16 above)
Comment (aborowski-msol) [reply samples/drivers/adc/src/main.c]: done

File: samples/drivers/adc/boards/nucleo_h743zi.overlay (hunk 10 above)
Comment (aborowski-msol) [reply samples/drivers/adc/boards/nucleo_h743zi.overlay]: not sure what that described originally, and what should be described instead

Comment (heinwessels) [reply samples/drivers/adc/boards/nucleo_h743zi.overlay]: Each node in the `dmas` is defined as follows, which is what all the numbers mean.
```
<node-name channel slot channel-config features>
```
The slot is `adc1_dma` I think, as defined in the reference manual in the DMA section, referring to a channel number or something. We need to be able to say what this number means, because it will influence our implementation as well.

File: drivers/adc/adc_stm32.c [hunk 18]
```diff
@@ -90,15 +96,15 @@ static const uint32_t table_resolution[] = {
...
+#define SMP_TIME(x, y)          LL_ADC_SAMPLINGTIME_##x##CYCLE##y
 
 /*
  * Conversion time in ADC cycles. Many values should have been 0.5 less,
  * but the adc api system currently does not support describing 'half cycles'.
  * So all half cycles are counted as one.
  */
 #if defined(CONFIG_SOC_SERIES_STM32F0X) || defined(CONFIG_SOC_SERIES_STM32F1X)
-static const uint16_t acq_time_tbl[8] = {2, 8, 14, 29, 42, 56, 72, 240};
+static const uint16_t acq_time_tbl[8] = { 2, 8, 14, 29, 42, 56, 72, 240 };
```
Comment (heinwessels) [inline drivers/adc/adc_stm32.c]: These changes seem odd to me. Even on upstream Zephyr there's no spaces there. Does the format fail if those spaces are removed?

Comment applies for similar cases below.

File: drivers/adc/adc_stm32.c [hunk 19]
```diff
@@ -471,39 +578,40 @@ static int start_read(const struct device *dev,
...
-		return -ENOTSUP;
-	}
-
 	data->buffer = sequence->buffer;
+	data->channel_count = 0;
+	data->samples_count = 0;
 
-	uint32_t channel = __LL_ADC_DECIMAL_NB_TO_CHANNEL(index);
+	for (uint8_t chSel = 0; chSel < STM32_CHANNEL_COUNT; chSel++) {
+		if (sequence->channels & (1 << chSel)) {
```
Comment (heinwessels) [inline drivers/adc/adc_stm32.c]:
```suggestion
		if (sequence->channels & BIT(chSel)) {
```

File: drivers/adc/adc_stm32.c [hunk 20]
```diff
@@ -650,12 +762,13 @@ static void adc_stm32_isr(const struct device *dev)
 	const struct adc_stm32_cfg *config =
 		(const struct adc_stm32_cfg *)dev->config;
 	ADC_TypeDef *adc = config->base;
+	int adcData = LL_ADC_REG_ReadConversionData32(adc);
```
Comment (heinwessels) [inline drivers/adc/adc_stm32.c]: We need to follow Zephyr convention here
```suggestion
	int adc_data = LL_ADC_REG_ReadConversionData32(adc);
```

Is there a reason why this is always read during an ISR? And not only when `data->buffer[data->samples_count++] =`? If so, then add a comment above the line saying "Read to clear the read flag" or something.

File: drivers/dma/dma_stm32.c [hunk 21]
```diff
@@ -132,15 +132,15 @@ static void dma_stm32_irq_handler(const struct device *dev, uint32_t id)
 
 #ifdef CONFIG_DMA_STM32_SHARED_IRQS
 
-#define HANDLE_IRQS(index)						       \
```
Comment (heinwessels) [inline drivers/dma/dma_stm32.c]: We're not chaning anything in this file, so I we should remove all changes to `dma_stm32.c` in this PR. This is to keep the PR clean.

File: drivers/dma/dma_stm32.c [hunk 22]
```diff
@@ -594,8 +595,11 @@ static int dma_stm32_init(const struct device *dev)
 	const struct dma_stm32_config *config = dev->config;
 	const struct device *clk = DEVICE_DT_GET(STM32_CLOCK_CONTROL_NODE);
 
+	/* This isn't called because dmamux is the master */
+	LOG_ERR("dma_stm32_init");
```
Comment (heinwessels) [inline drivers/dma/dma_stm32.c]: This also needs to be removed with the rest of the changes to this file.

File: drivers/dma/dmamux_stm32.c [hunk 23]
```diff
@@ -46,21 +46,21 @@ struct dmamux_stm32_config {
 	struct stm32_pclken pclken;
```
Comment (heinwessels) [inline drivers/dma/dmamux_stm32.c]: Same for this file. We're not making any changes to it and thus don't have to care about the formatting. So all changes should be removed.

File: samples/drivers/adc/boards/nucleo_h743zi.conf [hunk 24]
```diff
@@ -0,0 +1,9 @@
+CONFIG_LOG=y
+CONFIG_ADC_LOG_LEVEL_DBG=y
+CONFIG_DMA_LOG_LEVEL_DBG=y
+CONFIG_LOG_MODE_IMMEDIATE=y
```
Comment (heinwessels) [inline samples/drivers/adc/boards/nucleo_h743zi.conf]: Only used during debugging.
```suggestion
```

File: samples/drivers/adc/src/main.c [hunk 25]
```diff
@@ -118,5 +118,6 @@ void main(void)
 		printk("\n");
 
 		k_sleep(K_MSEC(1000));
+		break;
```
Comment (heinwessels) [inline samples/drivers/adc/src/main.c]: Rement debugging code.

File: samples/drivers/adc/src/main.c [hunk 26]
```diff
@@ -118,5 +118,6 @@ void main(void)
 		printk("\n");
```
Comment (heinwessels) [inline samples/drivers/adc/src/main.c]: Well done on getting it to run without any changes in this file!

File: drivers/adc/adc_stm32.c [hunk 27]
```diff
@@ -791,12 +905,12 @@ static int adc_stm32_channel_setup(const struct device *dev,
...
-		(__LL_ADC_CHANNEL_TO_DECIMAL_NB(ADC_CHANNEL_VREFINT) == channel_cfg->channel_id)) {
+	    (__LL_ADC_CHANNEL_TO_DECIMAL_NB(ADC_CHANNEL_VREFINT) == channel_cfg->channel_id)) {
 		adc_stm32_set_common_path(dev);
 	}
 
 	adc_stm32_setup_speed(dev, channel_cfg->channel_id,
-				  acq_time_index);
+			      acq_time_index);
 
 	LOG_DBG("Channel setup succeeded!");
```
Comment (heinwessels) [inline drivers/adc/adc_stm32.c]: This messages is printed for every channel and gets confusing.
```suggestion
	LOG_DBG("Channel %d setup succeeded!", channel_cfg->channel_id);
```

**Review by heinwessels — CHANGES_REQUESTED** [review]

File: drivers/adc/adc_stm32.c (hunk 18 above)
Comment (aborowski-msol) [reply drivers/adc/adc_stm32.c]: I would say this change was introduced by code formatting script. For sure I did not that manually.

File: drivers/adc/adc_stm32.c (hunk 19 above)
Comment (aborowski-msol) [reply drivers/adc/adc_stm32.c]: changed

File: drivers/adc/adc_stm32.c (hunk 20 above)
Comment (aborowski-msol) [reply drivers/adc/adc_stm32.c]: Just a leftover, as this was used just to log the value
fixed

File: drivers/adc/adc_stm32.c (hunk 27 above)
Comment (aborowski-msol) [reply drivers/adc/adc_stm32.c]: done

File: drivers/adc/adc_stm32.c (hunk 18 above)
Comment (heinwessels) [reply drivers/adc/adc_stm32.c]: That's exactly why I made this comment. As mentioned the script isn't perfect and shouldn't be taken as the truth. The upstream Zephyr doesn't have this change, and we intend to push these changes there.

File: samples/drivers/adc/boards/nucleo_h743zi.conf (hunk 24 above)
Comment (aborowski-msol) [reply samples/drivers/adc/boards/nucleo_h743zi.conf:4]: removed

File: drivers/dma/dmamux_stm32.c (hunk 23 above)
Comment (aborowski-msol) [reply drivers/dma/dmamux_stm32.c]: restored original file

File: drivers/dma/dma_stm32.c (hunk 22 above)
Comment (aborowski-msol) [reply drivers/dma/dma_stm32.c]: restored original file

File: drivers/dma/dma_stm32.c (hunk 21 above)
Comment (aborowski-msol) [reply drivers/dma/dma_stm32.c]: restored original file

File: drivers/adc/adc_stm32.c (hunk 18 above)
Comment (aborowski-msol) [reply drivers/adc/adc_stm32.c]: formatting restored

File: drivers/adc/adc_stm32.c [hunk 28]
```diff
@@ -233,6 +260,86 @@ struct adc_stm32_cfg {
...
+			 uint32_t channel, int status)
+{
+	/* arg directly holds the adc device */
+	struct adc_stm32_data *data = arg;
+
+	/* invalidate cache to make sure to get the adc data*/
+	SCB_InvalidateDCache_by_Addr(data->buffer, sizeof(uint16_t) * data->channel_count);
+	SCB_InvalidateDCache_by_Addr(data->repeat_buffer, sizeof(uint16_t) * data->channel_count);
+
+	data->samples_count = data->channel_count;
```
Comment (heinwessels) [inline drivers/adc/adc_stm32.c]: This line doesn't do anything.
```suggestion
```

File: drivers/adc/adc_stm32.c [hunk 29]
```diff
@@ -218,11 +238,18 @@ struct adc_stm32_data {
...
+	uint8_t samples_count;
 #if defined(CONFIG_SOC_SERIES_STM32F0X) || \
 	defined(CONFIG_SOC_SERIES_STM32G0X) || \
 	defined(CONFIG_SOC_SERIES_STM32L0X)
 	int8_t acq_time_index;
 #endif
+
+#ifdef CONFIG_ADC_STM32_DMA
+	struct k_sem status_sem;
+	volatile uint32_t status_flags;
```
Comment (heinwessels) [inline drivers/adc/adc_stm32.c]: This is also not used.
```suggestion
```

Comment (aborowski-msol) [reply drivers/adc/adc_stm32.c]: removed.

File: drivers/adc/adc_stm32.c (hunk 28 above)
Comment (aborowski-msol) [reply drivers/adc/adc_stm32.c]: it keeps the data consistent, so the sample count at the end of conversion should match channel_count

File: samples/drivers/adc/boards/nucleo_h743zi.conf (hunk 24 above)
Comment (aborowski-msol) [reply samples/drivers/adc/boards/nucleo_h743zi.conf]: changed to CONFIG_NO_OPTIMIZATIONS

Comment (heinwessels) [reply samples/drivers/adc/boards/nucleo_h743zi.conf]: We can't rely on `CONFIG_NO_OPTIMIZATIONS=y`. In some of our optima apps the application doesn't fit in FLASH without optimizations.

File: drivers/adc/adc_stm32.c [hunk 30]
```diff
@@ -651,11 +762,11 @@ static void adc_stm32_isr(const struct device *dev)
...
-	*data->buffer++ = LL_ADC_REG_ReadConversionData32(adc);
-
-	adc_context_on_sampling_done(&data->ctx, dev);
-
-	LOG_DBG("ISR triggered.");
+	if (data->samples_count == data->channel_count) {
+		adc_context_on_sampling_done(&data->ctx, dev);
+	} else {
+		data->buffer[data->samples_count++] = LL_ADC_REG_ReadConversionData32(adc);
+	}
```
Comment (heinwessels) [inline drivers/adc/adc_stm32.c]:
```suggestion
	/* Always read to clear the ISR flag */
	uint16_t read_adc_value = LL_ADC_REG_ReadConversionData32(adc);

	if (data->samples_count == data->channel_count) {
		adc_context_on_sampling_done(&data->ctx, dev);
	} else {
		data->buffer[data->samples_count++] = read_adc_value;
	}
```

File: drivers/adc/adc_stm32.c [hunk 31]
```diff
@@ -233,6 +259,88 @@ struct adc_stm32_cfg {
...
 };
 
+#ifdef CONFIG_ADC_STM32_DMA
+static void dma_callback(const struct device *dev, void *arg,
+			 uint32_t channel, int status)
+{
+	/* arg directly holds the adc device */
+	struct adc_stm32_data *data = arg;
+
+#if __DCACHE_PRESENT == 1
```
Comment (heinwessels) [inline drivers/adc/adc_stm32.c]: As mentioned in Slack, this check is already done inside `SCB_InvalidateDCache_by_Addr`. No need to add it again.

File: drivers/adc/adc_stm32.c (hunk 30 above)
Comment (aborowski-msol) [reply drivers/adc/adc_stm32.c]: fixed

File: samples/drivers/adc/src/main.c (hunk 25 above)
Comment (aborowski-msol) [reply samples/drivers/adc/src/main.c]: removed

File: samples/drivers/adc/boards/nucleo_h743zi.conf [hunk 32]
```diff
@@ -0,0 +1,8 @@
+CONFIG_LOG=y
+CONFIG_ADC_LOG_LEVEL_DBG=y
+CONFIG_DMA_LOG_LEVEL_DBG=y
```
Comment (heinwessels) [inline samples/drivers/adc/boards/nucleo_h743zi.conf]: I think this can be removed as well.

File: samples/drivers/adc/sample.yaml [hunk 33]
```diff
@@ -4,7 +4,7 @@ tests:
   sample.drivers.adc:
     tags: ADC
     depends_on: adc
-    platform_allow: nucleo_l073rz disco_l475_iot1 cc3220sf_launchxl
+    platform_allow: nucleo_l073rz nucleo_h732rz disco_l475_iot1 cc3220sf_launchxl
```
Comment (heinwessels) [inline samples/drivers/adc/sample.yaml]:
```suggestion
    platform_allow: nucleo_l073rz nucleo_h743zi disco_l475_iot1 cc3220sf_launchxl
```

File: drivers/adc/adc_stm32.c [hunk 34]
```diff
@@ -233,6 +260,96 @@ struct adc_stm32_cfg {
...
+			 &stream->dma_cfg);
+
+	/* the channel is the actual stream from 0 */
+	if (ret != 0) {
+		LOG_ERR("Problem setting up DMA %d", ret);
+		return ret;
+	}
+
+	/* invalidate cache to make sure to get the adc data*/
+	SCB_InvalidateDCache_by_Addr(sequence->buffer,
```
Comment (heinwessels) [inline drivers/adc/adc_stm32.c:326]: I think there should be a comment here saying that the buffer needs to be `__aligned(32)`.

File: drivers/adc/adc_stm32.c [hunk 35]
```diff
@@ -233,6 +260,96 @@ struct adc_stm32_cfg {
...
+	struct stream *stream = &data->dma;
+
+	blk_cfg = &stream->dma_blk_cfg;
+
+	/* prepare the block */
+	memset(blk_cfg, 0, sizeof(struct dma_block_config));
+	blk_cfg->block_size = sequence->buffer_size;
+
+	/* Source and destination */
+
```
Comment (heinwessels) [inline drivers/adc/adc_stm32.c]: Odd new line
```suggestion
```

**Review by heinwessels — COMMENTED** [review]
Nice! Just little cosmetics left.

Did you test if the code is reading the correct channels? And reading 16 channels at once?

File: drivers/adc/adc_stm32.c (hunk 31 above)
Comment (aborowski-msol) [reply drivers/adc/adc_stm32.c]: Removed

File: samples/drivers/adc/sample.yaml (hunk 33 above)
Comment (aborowski-msol) [reply samples/drivers/adc/sample.yaml]: changed

File: samples/drivers/adc/boards/nucleo_h743zi.conf (hunk 32 above)
Comment (aborowski-msol) [reply samples/drivers/adc/boards/nucleo_h743zi.conf]: I think only CONFIG_ADC_STM32_DMA=y is required here

File: drivers/adc/adc_stm32.c (hunk 34 above)
Comment (aborowski-msol) [reply drivers/adc/adc_stm32.c:326]: commented

File: drivers/adc/adc_stm32.c (hunk 35 above)
Comment (aborowski-msol) [reply drivers/adc/adc_stm32.c]: removed

**Comment by aborowski-msol** [comment]
> Nice! Just little cosmetics left.
> 
> Did you test if the code is reading the correct channels? And reading 16 channels at once?

I did the test with 4 channels, but not 16

Tested with shield: Uin1 - ADC1 CH3 and Rin2 - ADC1 CH18. Changing inputs corresponds to the changes of values displayed on the console. No other shield perpherals connected to ADC1

**Review by heinwessels — APPROVED** [review]
